storage/src/adb/store/batcher.rs

```diff
@@ -0,0 +1,452 @@
+    V: Codec + Clone,
+    T: Translator,
+    D: Db<E, K, V, T>,
+> {
+    db: D,
+    updates: HashMap<K, UpdateOrDelete<V>>,
+    _phantom: PhantomData<(E, T)>,
+}
+
+impl<
+        E: Storage + Clock + Metrics,
+        K: Array,
+        V: Codec + Clone,
+        T: Translator,
+        D: Db<E, K, V, T>,
+    > Batcher<E, K, V, T, D>
+{
+    /// Create a new batcher wrapping the given database.
+    pub fn new(db: D) -> Self {
+        Self {
+            db,
+            updates: HashMap::new(),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Get a reference to the underlying database.
+    pub fn db(&self) -> &D {
+        &self.db
+    }
+
+    /// Take the underlying database from the batcher. Any cached updates will be applied before returning.
+    pub async fn take(mut self) -> Result<D, Error> {
+        self.apply_updates().await?;
+
+        Ok(self.db)
+    }
+
+    /// Delete the value assigned to `key` in the database, if any. This version is more efficient
+    /// than regular `delete` because it doesn't provide a return value indicating whether the key
+    /// was already deleted.
+    pub fn delete_unchecked(&mut self, key: K) {
+        if let Some(update_or_delete) = self.updates.get_mut(&key) {
+            match update_or_delete {
+                UpdateOrDelete::Update(_) => {
+                    *update_or_delete = UpdateOrDelete::Delete;
+                }
+                UpdateOrDelete::Delete => {}
+            }
+            return;
+        }
+
+        self.updates.insert(key, UpdateOrDelete::Delete);
+    }
+
+    /// Applies any cached updates to the underlying db via the batch update method without applying
+    /// a commit operation.
+    ///
+    /// # Errors
+    ///
+    /// - Propagates any underlying database errors. Because these errors are generally unrecoverable,
+    ///   no effort is made to preserve batcher state should one occur.
+    pub async fn apply_updates(&mut self) -> Result<(), Error> {
+        let updates = std::mem::take(&mut self.updates);
+        let updates_iter =
+            updates
+                .iter()
+                .filter_map(|(key, update_or_delete)| match update_or_delete {
+                    UpdateOrDelete::Update(value) => Some((key.clone(), value.clone())),
+                    UpdateOrDelete::Delete => None,
+                });
+        let deletes_iter =
+            updates
+                .iter()
+                .filter_map(|(key, update_or_delete)| match update_or_delete {
+                    UpdateOrDelete::Update(_) => None,
+                    UpdateOrDelete::Delete => Some(key.clone()),
+                });
+
+        self.db.batch_update(updates_iter, deletes_iter).await
+    }
+}
+
+impl<
+        E: Storage + Clock + Metrics,
+        K: Array,
+        V: Codec + Clone,
+        T: Translator,
+        D: Db<E, K, V, T>,
+    > Db<E, K, V, T> for Batcher<E, K, V, T, D>
+{
+    fn op_count(&self) -> Location {
```

Comment:
I wonder whether users might mistakenly expect `op_count` to include updates/deletions which are in `updates` but haven't yet been flushed to the underlying database.

Reply:
I have changed the batcher's op_count apply updates before returning op_count.  This isn't exactly ideal but the alternative would be potential non-monotonicity in op_count if cached updates are later deleted, which might be weird.

Let's revisit whether to adopt a first-class Batch abstraction? Seems like it might be a much bigger change which I'm not quite ready to pursue.

Reply:
Bah, unfortunately this will require making op_count async.